addons/stock/models/stock_move.py

```diff
@@ -1406,12 +1406,12 @@ def _assign_picking_values(self, picking):
         if any(picking.origin != m.origin for m in self):
             current_origins = set(picking.origin.split(',') + [False]) if picking.origin else {False}
             vals['origin'] = picking.origin
-            for move in self:
-                if move.origin not in current_origins:
-                    if not vals['origin']:
-                        vals['origin'] = move.origin
-                    else:
-                        vals['origin'] += f',{move.origin}'
+            moves = self.filtered(lambda move: move.origin not in current_origins and not current_origins.add(move.origin))
+            for move in moves:
+                if not vals['origin']:
+                    vals['origin'] = move.origin
+                else:
+                    vals['origin'] += f',{move.origin}'
```

Comment:
I think this would be easier to read if it was written as a simple loop. Right now, the filter lambda modifies 
`current_origins` as a side effect, which is unexpected and hard to debug.